velox/functions/lib/DateTimeFormatter.cpp

```diff
@@ -927,9 +928,92 @@ void parseFromPattern(
+      case DateTimeFormatSpecifier::ERA:
+      case DateTimeFormatSpecifier::HALFDAY_OF_DAY:
+        // Fixed size.
+        size += 2;
+        break;
+      case DateTimeFormatSpecifier::YEAR_OF_ERA:
+        // Timestamp is in [-32767-01-01, 32767-12-31] range.
+        size += std::max((int)token.pattern.minRepresentDigits, 6);
+        break;
+      case DateTimeFormatSpecifier::CENTURY_OF_ERA:
+      case DateTimeFormatSpecifier::DAY_OF_WEEK_0_BASED:
+      case DateTimeFormatSpecifier::DAY_OF_WEEK_1_BASED:
+        size += std::max((int)token.pattern.minRepresentDigits, 1);
+        break;
+      case DateTimeFormatSpecifier::DAY_OF_WEEK_TEXT:
+      case DateTimeFormatSpecifier::MONTH_OF_YEAR_TEXT:
+        // 9 is the max size of elements in weekdaysFull or monthsFull.
+        size = token.pattern.minRepresentDigits <= 3 ? size + 3 : size + 9;
+        break;
+      case DateTimeFormatSpecifier::YEAR:
+        // Timestamp is in [-32767-01-01, 32767-12-31] range.
+        size = token.pattern.minRepresentDigits == 2
+            ? size + 2
+            : size + std::max((int)token.pattern.minRepresentDigits, 6);
+        break;
+      case DateTimeFormatSpecifier::DAY_OF_YEAR:
+        size += std::max((int)token.pattern.minRepresentDigits, 3);
+        break;
+      case DateTimeFormatSpecifier::MONTH_OF_YEAR:
+      case DateTimeFormatSpecifier::DAY_OF_MONTH:
+      case DateTimeFormatSpecifier::HOUR_OF_HALFDAY:
+      case DateTimeFormatSpecifier::CLOCK_HOUR_OF_HALFDAY:
+      case DateTimeFormatSpecifier::HOUR_OF_DAY:
+      case DateTimeFormatSpecifier::CLOCK_HOUR_OF_DAY:
+      case DateTimeFormatSpecifier::MINUTE_OF_HOUR:
+      case DateTimeFormatSpecifier::SECOND_OF_MINUTE:
+        size += std::max((int)token.pattern.minRepresentDigits, 2);
+        break;
+      case DateTimeFormatSpecifier::FRACTION_OF_SECOND:
+        size += token.pattern.minRepresentDigits;
+        break;
+      case DateTimeFormatSpecifier::TIMEZONE:
+        size += std::max(
+            token.pattern.minRepresentDigits, timezone->name().length());
+        break;
+      // Not supported.
+      case DateTimeFormatSpecifier::TIMEZONE_OFFSET_ID:
+      case DateTimeFormatSpecifier::WEEK_YEAR:
+      case DateTimeFormatSpecifier::WEEK_OF_WEEK_YEAR:
+      default:
+        VELOX_UNSUPPORTED(
+            "format is not supported for specifier {}",
+            token.pattern.specifier);
+    }
+  }
+  return size;
+}
+
 std::string DateTimeFormatter::format(
     const Timestamp& timestamp,
     const date::time_zone* timezone) const {
+  const auto maxSize = evaluateMaxResultSize(timezone);
+  std::string result(maxSize, '\0');
+  auto actualSize =
+      format(std::string_view(result.data(), maxSize), timestamp, timezone);
+  result.resize(maxSize);
+  return result;
+}
+
+// Result is pre-allocated according to evaluateMaxResultSize.
+// Returns actual size.
+int32_t DateTimeFormatter::format(
+    const std::string_view& result,
```

Comment:
this API is strange; std::string_view should be passed by value, but here I think we should use char* and make it last argument (input args first, output last).